controllers/clusters/postgresql_controller.go

```diff
@@ -492,11 +493,205 @@ func (r *PostgreSQLReconciler) handleUpdateCluster(
+	pg *v1beta1.PostgreSQL,
+	resourceKind string,
+	oldRefs, newRefs []*v1beta1.NamespacedName,
+) {
+	ctx := context.TODO()
+	l := log.FromContext(ctx)
+
+	for _, ref := range newRefs {
+		var exist bool
+		for _, oldRef := range oldRefs {
+			if *ref == *oldRef {
+				exist = true
+				break
+			}
+		}
+
+		if exist {
+			continue
+		}
+
+		err := r.handleCreateResource(ctx, l, resourceKind, ref, pg)
+		if err != nil {
+			l.Error(err, "Cannot create clusterresource", "resource kind", resourceKind, "namespace and name", ref)
+			r.EventRecorder.Eventf(pg, models.Warning, models.CreatingEvent,
+				"Cannot create resource. Reason: %v", err)
+		}
+		oldRefs = append(oldRefs, ref)
+	}
+	for _, oldRef := range oldRefs {
+		var exist bool
+		for _, ref := range newRefs {
+			if *oldRef == *ref {
+				exist = true
+				break
+			}
+		}
+
+		if exist {
+			continue
+		}
+
+		err := r.handleDeleteResource(ctx, l, resourceKind, oldRef)
+		if err != nil {
+			l.Error(err, "Cannot delete clusterresource", "resource kind", resourceKind, "namespace and name", oldRef)
+			r.EventRecorder.Eventf(pg, models.Warning, models.DeletingEvent,
+				"Cannot delete resource. Reason: %v", err)
+		}
+	}
+}
+
+func (r *PostgreSQLReconciler) handleCreateResource(
+	ctx context.Context,
+	l logr.Logger,
+	kind string,
+	ref *v1beta1.NamespacedName,
+	pg *v1beta1.PostgreSQL,
+) error {
+	req := types.NamespacedName{
+		Namespace: ref.Namespace,
+		Name:      ref.Name,
+	}
+
+	var resource clusterresources.Object
+	var isCDC bool
+
+	switch kind {
+	case models.ClusterbackupRef:
+		resource = &clusterresourcesv1beta1.ClusterBackup{}
+	case models.ClusterNetworkFirewallRuleref:
+		resource = &clusterresourcesv1beta1.ClusterNetworkFirewallRule{}
+	case models.AWSVPCPeeringRef:
+		resource = &clusterresourcesv1beta1.AWSVPCPeering{}
+		isCDC = true
+	case models.AWSSecurityGroupFirewallRuleRef:
+		resource = &clusterresourcesv1beta1.AWSSecurityGroupFirewallRule{}
+	case models.ExclusionWindowRef:
+		resource = &clusterresourcesv1beta1.ExclusionWindow{}
+	case models.GCPVPCPeeringRef:
+		resource = &clusterresourcesv1beta1.GCPVPCPeering{}
+		isCDC = true
+	case models.AzureVNetPeeringRef:
+		resource = &clusterresourcesv1beta1.AzureVNetPeering{}
+		isCDC = true
+	default:
+		l.Info("Provided reference to resource that is not supported", "kind", kind)
+		return nil
+	}
+
+	err := r.Get(ctx, req, resource)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			l.Error(err, "Cannot create a cluster resource. The resource is not found", "request", req)
+			return err
+		}
+		l.Error(err, "Cannot get cluster resource", "Resource", resource)
+		return err
+	}
+
+	patch := resource.NewPatch()
+
+	if isCDC {
+		resource.AttachToCluster(pg.Status.DataCentres[0].ID)
+	} else {
+		resource.AttachToCluster(pg.Status.ID)
+	}
+
+	err = r.Status().Patch(ctx, resource, patch)
+	if err != nil {
+		return err
+	}
+
+	l.Info("PostgreSQL clusterresource was patched",
+		"Resource name", ref.Name,
```

Comment:
I think you should log and namespace too
```suggestion
		"reference", ref,
```

Reply:
refactored